rules/REQUEST-903.9011-DOKUWIKI-EXCLUSION-RULES.conf

```diff
@@ -0,0 +1,275 @@
+    pass,\
+    skipAfter:END-DOKUWIKI"
+
+
+
+#
+# -=[ Dokuwiki Front-End ]=-
+#
+
+# Note on files specified:
+# /doku.php: shows pages, saves, edits, admin
+# /lib/exe/ajax.php: autosave, uploads
+
+
+# Allow pages to be edited, and ajax to save drafts.
+#
+# ARGS 'wikitext', 'suffix', and 'prefix' must allow the same things,
+# as the page (in part or whole) is passed via 'suffix/prefix' at times.
+#  attack-protocol (921110-921160/920230): Allows odd characters on the page.
+#  CRS: (still need attack-protocol specified.)
+#  attack-injection-php (930000-933999): Allows code on page.
+#  attack-sqli (940000-942999): Allows SQL expressions on page.
+#
+# Others:
+#  930100-930110;REQUEST_BODY: if there's  a /../ in the text.
+#
+# ARGS:summary (the text in the 'summary' box on page edits.):
+#  Allowing 930120-930130 lets user save summaries with
+#  system file names. This should not be needed in normal
+#  use. But leaving a note here of how to allow in rule below:
+#    ctl:ruleRemoveTargetByID=930120;ARGS:summary
+#    ctl:ruleRemoveTargetByID=930130;ARGS:summary
+#
+# Also, can't specify:
+#   SecRule ARGS:do "@streq edit" \
+#   SecRule REQUEST_FILENAME "@endsWith /lib/exe/ajax.php"\
+# because at times the do=edit can get dropped, so if we use
+# above the edit will get blocked when the page is saved.
+
+# Hint: those using .htaccess rewrites can remove/replace
+# this first 'SecRule...' line with 'SecAction \' (unsupported).
+
+SecRule REQUEST_FILENAME "@rx (/doku.php|/lib/exe/ajax.php)$" \
+    "id:9011100,\
+    phase:2,\
+    t:none,\
+    nolog,\
+    pass,\
+    chain"
+    SecRule REQUEST_METHOD "@streq POST" \
+        "t:none,\
+        chain"
+    SecRule REQUEST_COOKIES:/S?DW[a-f0-9]+/ '@rx ^[%a-zA-Z0-9_-]+' \
+        "t:none,\
+        ctl:ruleRemoveTargetByTag=attack-protocol;ARGS:wikitext,\
+        ctl:ruleRemoveTargetByTag=CRS;ARGS:wikitext,\
+        ctl:ruleRemoveTargetByTag=attack-protocol;ARGS:suffix,\
+        ctl:ruleRemoveTargetByTag=CRS;ARGS:suffix,\
+        ctl:ruleRemoveTargetByTag=attack-protocol;ARGS:prefix,\
+        ctl:ruleRemoveTargetByTag=CRS;ARGS:prefix,\
+        ctl:ruleRemoveTargetByID=930100-930110;REQUEST_BODY"
+
+
+# Allow it to upload files. But check for cookies just to make sure.
+
+SecRule REQUEST_FILENAME "@endsWith /lib/exe/ajax.php"\
+    "id:9011110,\
+    phase:2,\
+    t:none,\
+    pass,\
+    nolog,\
+    noauditlog,\
+    chain"
+    SecRule REQUEST_METHOD "@streq POST" \
+        "t:none,\
+        chain"
+    SecRule REQUEST_COOKIES:/S?DW[a-f0-9]+/ '@rx ^[%a-zA-Z0-9_-]+' \
+        "t:none,\
+        setvar:'tx.allowed_request_content_type=%{tx.allowed_request_content_type}|application/octet-stream'"
+
+
+# If it's a large upload, turn off scanning
+
+SecRule REQUEST_FILENAME "@endsWith /lib/exe/ajax.php"\
+    "id:9011120,\
+    phase:2,\
+    t:none,\
+    pass,\
+    nolog,\
+    noauditlog,\
+    chain"
+    SecRule REQUEST_HEADERS:Content-Length "@gt 31486341" \
```

Comment:
Are you choosing this number for any particular reason?

Reply:
No. Not anything definite beyond just seeming like a good balance for performance. But starting to think I should just comment it out at the top, for the config, if at all.

Reply:
I was actually not referring to removing that rule, but from my experience even running CRS on anything below 32MB seems rather problematic and can cause FPs and long latency/high CPU. But having the default which the user should set anyway should work fine.

Reply:
Well, that's what I tried before, setting the 'crs_exclusions_dokuwiki_scan_upload=31486341' variable that would respond to a rule setting the upload scan limit, but I couldn't get it to work. It was as if it was being ignored. Is that what you are referring to? I don't think it would be good to hard code it in, as if someone needed to modify it, they'd have to do so each time it was updated.

Or just set crs_exclusions_dokuwiki_scan_upload=0|1 for if uploaded files are scanned or not. And then a simple rule to turn off scanning for uploads.

Reply:
Option C: Put a note at the top that the user can add to crs-setup.conf or website config.

```
# If you want to relax the upload restrictions of file types,
# see rule 900240. For Dokuwiki you can limit the exception
# to the ajax.php file:
#
# SecRule REQUEST_FILENAME "@endsWith /lib/exe/ajax.php"\
#    "id:9011120,\
#    phase:1,\
#    t:none,\
#    nolog,\
#    pass,\
#    chain"
#    SecRule REQUEST_COOKIES:/S?DW[a-f0-9]+/ '@rx ^[%a-zA-Z0-9_-]+' \
#      "t:none,\
#      tx.restricted_extensions='.bak/ .config/ .conf/ .db/ .ini/ .log/ .old/ .pass/ .pdb/ .sql/'"
#
# If you have have performance/latency issues on large uploads, 
# you can try disabling scanning of uploaded files (may open security
# holes). Here's an example for crs-setup.conf or your website's config 
# that will only make the exception for uploads (ajax.php)
#
# SecRule REQUEST_FILENAME "@endsWith /lib/exe/ajax.php"\
#    "id:9011121,\
#    phase:1,\
#    t:none,\
#    pass,\
#    nolog,\
#    noauditlog,\
#    chain"
#    SecRule REQUEST_HEADERS:Content-Length "@gt 31486341" \
#        "t:none,\
#        ctl:requestBodyAccess=Off"
#
# And to increase the upload size for only uploaded files (512MB):
#
# SecRule REQUEST_FILENAME "@endsWith /lib/exe/ajax.php"\
#    "id:9011122,\
#    phase:1,\
#    t:none,\
#    pass,\
#    nolog,\
#    noauditlog,\
#    ctl:requestBodyLimit=536870912"
```